### ISSUE TITLE
Usage in Angular-Cli fails with CommonJS Error

### ISSUE THREAD

**Issue by flamusdiu** [jonschlinkert/markdown-toc#102]
I have following function:
```
createToc(doc) {
  const options = {}
   const toc = require('markdown-toc');     
        return Promise.resolve().then(() => {
            return new Remarkable()
                .use(toc.plugin(options))
                .render(doc);
        });
}
```
When building the  application, I get the following message:
```
WARNING in ./node_modules/markdown-toc/lib/utils.js
8:34-41 Critical dependency: require function is used in a way in which dependencies cannot be statically extracted
```
From [webpack/webpack#2675](https://github.com/webpack/webpack/issues/2675), I guess the issue might be here: [utils.js#L26](https://github.com/jonschlinkert/markdown-toc/blob/3c6eb6a5c67707eeced3f0822e9a3a160d4b7321/lib/utils.js#L26)

**Comment by flamusdiu** [comment]
Umm, I would have to use [doowb/unlazy-loader](https://github.com/doowb/unlazy-loader) to get this to work; however, Angular CLI doesn't seem to provided anyway to inject the plugin that I have found without configuring the whole webpack stack myself.

**Comment by doowb** [comment]
I meant to mention the unlazy loader earlier... I didn't know it was hard to add webpack loaders to angular-cli.

I'll look into this a little more later today.

**Comment by flamusdiu** [comment]
Yeah, you can run `ng eject` but that makes the webpack config static and you lose the cli stuff. Thanks for the response. You can see more from [angular/angular-cli#1665](https://github.com/angular/angular-cli/issues/1656)

**Comment by doowb** [comment]
@flamusdiu I don't know much about `angular-cli` but it looks like from the issue they're not going to allow customizing webpack configs.

I don't have the time right now, but if someone else wants to do a PR, we could browserify the source into a distributable file and set the `browser` property in `package.json` to point to that file. I'm pretty sure webpack supports the `browser` property now, so that should work for `angular-cli` builds too.

**Comment by flamusdiu** [comment]
`webpack` does but not sure if `angular-cli` will use it. You would need to set it the `angular-cli` file. I guess I'll need to look at how to do that.

**Comment by flamusdiu** [comment]
I forked it and did a quick fix for the moment: [flamusdiu/markdown-toc](https://github.com/flamusdiu/markdown-toc)

Also, you have to use [gray-matter-webpack-build](https://www.npmjs.com/package/gray-matter-webpack-build) as well as change the require statements.  Not sure if there is a good way to change the dependencies when installing.

**Comment by flamusdiu** [comment]
Hell, I hit the same problem as markdown-toc#82

**Comment by flamusdiu** [comment]
Okay, so #82 is caused because `angular-cli` uses `webpack` which uses `querystring-es3`. When webpack compiles, it uses this resolution which does not have the `.escape` function.

**Comment by flamusdiu** [comment]
Oooo ...  Should get fixed with this -> https://github.com/webpack/node-libs-browser/pull/70

**Comment by flamusdiu** [comment]
@doowd if you don't want to wait, could use the solution in the comment https://github.com/mhart/aws4/issues/46#issuecomment-313792048 changing the function. It might more limited by, though.

**Comment by doowb** [comment]
@flamusdiu thanks for digging in more and looking for a solution.

On your fork, if you try some of the following things that end up working with angular-cli, I'll review and consider a PR:

- Instead of using `gray-matter-webpack-build`, try upgrading `gray-matter` to the latest version. There were some changes made because of webpack there.
- In the `markdown-toc` `package.json`, add `querystring-es3` as a dependency and add the following property:
  ```json
  "browser": {
    "querystring": "querystring-es3"
  },
  ```

Along with your changes to require in the utils directly, I think these would make working with webpack and angular-cli easier.

**Comment by flamusdiu** [comment]
I'll try this in the morning

**Comment by flamusdiu** [comment]
@doowb I am still getting `querystring.escape is not a function`. You can check my branch here: https://github.com/flamusdiu/markdown-toc/tree/try-changes-1

I'll look at it again when I get off work.

**Comment by HereChen** [comment]
I have the same problem in a vue-cli project (webpack).

```
import markdownToc from 'markdown-toc';
...
markdownToc.Insert(markdown);
...
```

```
 warning  in ./~/markdown-toc/lib/utils.js

8:34-41 Critical dependency: require function is used in a way in which dependencies cannot be statically extracted
```

**Comment by Aman-Jojra** [comment]
I am facing a similar issue: [Stackoveflow](https://stackoverflow.com/questions/51319209/critical-dependency-require-function-is-used-in-a-way-in-which-dependencies-can)

has anyone faced this issue before?
